include/nbl/system/ISystem.h

```diff
@@ -0,0 +1,244 @@
+#ifndef __NBL_I_SYSTEM_H_INCLUDED__
+#define __NBL_I_SYSTEM_H_INCLUDED__
+
+#include <variant>
+#include "nbl/core/IReferenceCounted.h"
+#include "nbl/system/ISkippableAsyncQueueDispatcher.h"
+#include "nbl/system/IFileArchive.h"
+#include "nbl/system/IWriteFile.h"
+
+namespace nbl {
+namespace system
+{
+
+class ISystem final : public core::IReferenceCounted
+{
+public:
+    class ISystemCaller : core::IReferenceCounted
+    {
+    protected:
+        virtual ~ISystemCaller() = default;
+
+    public:
+        virtual core::smart_refctd_ptr<IReadFile> createReadFile(ISystem* sys, const std::filesystem::path& filename) = 0;
+        virtual core::smart_refctd_ptr<IWriteFile> createWriteFile(ISystem* sys, const std::filesystem::path& filename) = 0;
+        virtual uint32_t read(IReadFile* file, void* buffer, size_t offset, size_t size) = 0;
+        virtual uint32_t write(IWriteFile* file, const void* buffer, size_t offset, size_t size) = 0;
```

Comment:
maybe make it a protected method of the file?

Reply:
nah, i want everything OS-specific to live in ISystemCaller. Or maybe you meant something different

Reply:
yeah I think you've convinced me to have the methods here, and not have any methods in the file.

resolved.